go/vt/vtctl/reparentutil/emergency_reparenter.go

```diff
@@ -761,3 +772,184 @@ func (erp *EmergencyReparenter) filterValidCandidates(validTablets []*topodatapb
 	}
 	return restrictedValidTablets, nil
 }
+
+// findErrantGTIDs tries to find errant GTIDs for the valid candidates and returns the updated list of valid candidates.
+func (erp *EmergencyReparenter) findErrantGTIDs(
+	ctx context.Context,
+	validCandidates map[string]replication.Position,
+	statusMap map[string]*replicationdatapb.StopReplicationStatus,
+	tabletMap map[string]*topo.TabletInfo,
+	waitReplicasTimeout time.Duration,
+) (map[string]replication.Position, error) {
+	// First we need to collect the reparent journal length for all the candidates.
+	// This will tell us, which of the tablets are severly lagged, and haven't even seen all the primary promotions.
+	// These tablets cannot be trusted for errant GTID detection.
+	reparentJournalLen, err := erp.gatherReparenJournalInfo(ctx, validCandidates, tabletMap, waitReplicasTimeout)
+	if err != nil {
+		// If we run into an error finding the reparent journal lengths, we use the legacy errant GTID detection logic.
+		return erp.legacyFindErrantGTIDS(validCandidates, statusMap)
+	}
+
+	// Find the maximum length of the reparent journal among all the candidates.
+	var maxLen int
+	for _, length := range reparentJournalLen {
+		maxLen = max(maxLen, length)
+	}
+
+	// Find the candidates with the maximum length of the reparent journal.
+	var maxLenCandidates []string
+	for alias, length := range reparentJournalLen {
+		if length == maxLen {
+			maxLenCandidates = append(maxLenCandidates, alias)
+		}
+	}
+
+	// We use all the candidates with the maximum length of the reparent journal to find the errant GTIDs amongst them.
+	var maxLenPositions []replication.Position
+	updatedValidCandidates := make(map[string]replication.Position)
+	for _, candidate := range maxLenCandidates {
+		status, ok := statusMap[candidate]
+		if !ok {
+			// If the tablet is not in the status map, and has the maximum length of the reparent journal,
+			// then it is probably the latest primary and we don't need to run any errant GTID detection on it!
```

Comment:
Yes, we can't run errant GTID detection on the primary, because it can have writes that no other tablet does. Like if the semi-sync replica was down that asked those writes. We don't know the server UUID of the tablet. We can alternately start storing it as well, and use its own server UUID to discount the writes.

Reply:
I am rewording the comment however to give a little more detail.

Reply:
Here is the new comment
```
// If the tablet is not in the status map, and has the maximum length of the reparent journal,
			// then it should be the latest primary and we don't need to run any errant GTID detection on it!
			// There is a very unlikely niche case that can happen where we see two tablets report themselves as having
			// the maximum reparent journal length and also be primaries. Here is the outline of it -
			// 1. Tablet A is the primary and reparent journal length is 3.
			// 2. It gets network partitioned, and we promote tablet B as the new primary.
			// 3. tablet B gets network partitioned before it has written to the reparent journal, and a new ERS call ensues.
			// 4. During this ERS call, both A and B are seen online. They would both report being primary tablets with the same reparent journal length.
			// Even in this case, the best we can do is not run errant GTID detection on either, and let the split brain detection code
			// deal with it, if A in fact has errant GTIDs.
```

Reply:
Looks good.